### PR TITLE
Aem as a Cloud Service incompatibility banner

### PR DESCRIPTION
**Pull request by davidjgonzalez**
Adds a banner to the feature docs for ACS Commons if frontmatter has

`tags: aemcs-incompatible`

Took a first pass at tagging obvious candidates.

image

### CODE REVIEW

**Comment by badvision** [comment]
Versioned client libs are incompatible?  That's a bummer -- is it because they persist outside of /content or something?

**Comment by davidjgonzalez** [comment]
@badvision not sure the details, but there is an OOTB replacement for them that is "fully AEM as a Cloud Service" compatible.

This message may be a bit heavy handed - so wording suggestions are welcome. Basically, anything tagged with this should not be used on AEM as a Cloud Service, the feature:

* May simply not work
* May break at any given point in the future since it relies on private/internal/super-deprecated functionality
* Has been superseded by another OOTB feature/approach
* May cause problems at times, but other times works fine; ie. eventually youll start causing issues for yourself if you use them.

Basically.. "dont use this on AEM as a Cloud Service" ... not sure the best wording for it. @ireasor thought the messaging in the PR might be a bit harsh/alarmist :P

**Comment by davidjgonzalez** [comment]
@badvision im sure you noticed i maked up a handful of other feature docs i know are incompatible and/or bad practices in AEM as a Cloud Service.. im sure there are more. A big one is pulling binaries into memory and writing them to AEM via Sling/AEM/JCR Java APIs.. the correct pattern is to do the multi-request POST operation where you register the binary in AEM (segment store) but actually PUSH the binary directly into the blobstore (if i understand it correctly)... basically, we dont support "pulling" assets/binaries into AEM, but rather it must be a push from another system.

**Comment by badvision** [comment]
I think it is certainly safe to say "This is not confirmed to work on AEM CS, use at your own risk" or something like that.  Anything we know will be for sure problematic we might want a red blinking banner... or do the right thing and just EOL some stuff.

**Comment by kwin** [comment]
There should be a link in that box explaining the "why". I guess in this case it is about violating content-classifications?

**Comment by badvision** [comment]
> There should be a link in that box explaining the "why". I guess in this case it is about violating content-classifications?

In many cases, that's what we should expect the root cause to be; but some features such as those which rely on ClassicUI are incompatible because ClassicUI is not available anymore -- that day was known to be coming since AEM 6.1, really.   The only feature I'm unsure about is the on-deploy publisher scripting, because publisher architecture was changed in AEMCS, and so that one feature I would caution people to use at their own risk and test it heavily before going live with it.

File: _acs-aem-commons/features/workflow-processes/assets-rendition-matter/index.md [hunk 1]
```diff
@@ -5,7 +5,7 @@ description: Matte your assets!
 sub-feature: true
 date: 2013-10-02
 initial-release: 1.2.0
----
+tags: aemcs-incompatible
```
Comment (davidjgonzalez) [inline _acs-aem-commons/features/workflow-processes/assets-rendition-matter/index.md]:
```suggestion
----
```

**Comment by justinedelson** [comment]
caveat -- bit of a peanut gallery comment here...

I think it'd be useful to differentiate between "conceptually incompatible" and "implementation incompatible". And if the problem is that the current implementation is incompatible, maybe link to a github issue (labeled "help-wanted")

**Comment by davidjgonzalez** [comment]
@justinedelson  can you give examples of features that would be **conceptually incompatible** vs. **implementation incompatible**?

To me, the guidance is ideally binary:

1) It's OK to use this on AEM as a Cloud Service
2) It's NOT OK to use this on AEM as a Cloud Service

If the conceptually vs implementation gives us some leeway, we could do that (assuming we have clear definitions). What's something that conceptually incompatible that we'd still feel comfortable not giving it an explicit "dont use"? .. ClassicUI stuff?


@kwin If we want to dive into "Why" (which is certainly nice, just more work for us)  I think we'll need to elaborate per-feature (some might have common reasons; but thats ~4-5 why's across the things I marked in this PR)

**Comment by justinedelson** [comment]
@davidjgonzalez it's more that for the conceptually compatible features (but cases where the implementation isn't compatible) there should be a request for help to change the implementation. In other words, I'm working backwards a bit -- some of these (I assume) are opportunities to drive contribution.

But totally possible that nothing here is "just" an implementation gap and there's no opportunity.

**Comment by kwin** [comment]
I think starting to collect the possible reasons in a central place is a good start. We should only differentiate between:
1. needs work
2. will never work in the current form

Only for the latter we need the "why"!

**Comment by joerghoh** [comment]
@kwin I guess even in the case "needs work" we should outline what's the problem with the existing implementation.

**Comment by davidjgonzalez** [comment]
@kwin That seems reasonable to me - with the overarching guidance of for BOTH being, don't use this on AEM as a Cloud Service *today* (since it either needs 1) updating to work on AEM as a Cloud Service or 2) it will never work on AEM as a CLoud Service give AEM's capabilities/roadmap as we know them today.

We can of course always remove/update "compatibility status" as we AEM as a Cloud Service evolves and/or we learn different ways to adjust to new paradigms.

...and then +1 to a "catalog" of "Whys".

WDYT?

**Comment by shsteimer** [comment]
@badvision @davidjgonzalez I'm wondering if it would make sense to always link this with an associated issue.  That way the banner could be somewhat generic "this won't work with AEMCS, see here for more info", with an issue link that provides further context on why it is incompatible, and what if anything can be done to make it compatible, what to use instead, workaround, etc.  In cases where it is "conceptually incompatible" the issue could be used to track deprecating and eventually removing the feature?

**Comment by ireasor** [comment]
Considering this from the point of view of how it would be implemented, maybe we should have an additional field on the md file for open-issues or known-issues that can point to known or open issues.  We can implement the basic messaging that @davidjgonzalez already has and then add an issue to this field that contains either an issue to deprecate and remove the feature down the road (conceptually incompatible) or to fix it to work on AEMCS.  The nice part of this approach is that we could use it for any issue, even those that are unrelated to AEMCS per se, to help raise visibility of known issues and encourage contribution.

**Comment by davidjgonzalez** [comment]
Merging for now, just do we have something helping guide people in terms of what could/should be used. We can incorporate enhancements in new PR (and/or if someone wants to change the wording on this)

**Comment by cpilsworth** [comment]
>  do we have something helping guide people in terms of what could/should be used.
is it worth mentioning the asset bulk ingestion feature here now?
* https://experienceleague.adobe.com/docs/adobe-developers-live-events/assets/asset-bulk-ingestion.pdf?lang=en
* https://experienceleague.adobe.com/docs/experience-manager-cloud-service/assets/manage/add-assets.html?lang=en#asset-bulk-ingestor
